tools/depcheck/pkg/cmd/trace/trace.go

```diff
@@ -0,0 +1,431 @@
+	cmd.Flags().StringVarP(&flags.OutputFormat, "output", "o", "", "output generated dependency graph in specified format. One of: dot.")
+	cmd.Flags().StringSliceVar(&flags.Subroots, "subroot", flags.Subroots, "root packages of sub-trees contained within trees initially specified via --root. A set of transitive dependencies unique to these trees relative to the rest of the dependency graph is returned.")
+	cmd.Flags().BoolVar(&flags.Shared, "shared", flags.Shared, "indicates whether to include the joint set of dependencies between --subroot trees and the rest of the graph in the final analysis.")
+	cmd.Flags().IntVar(&flags.Depth, "depth", -1, "if provided with a non-negative integer, blames packages at the provided depth-level responsible for bringing in any --shared dependencies.")
+	return cmd
+}
+
+func (o *TraceImportsOpts) Complete() error {
+	o.Roots = expandRecursePackages(o.Roots)
+	return nil
+}
+
+// expandRecursePackages receives a list of root packages specified
+// via --root, and ensures that each path ends in an ellipsis (...).
+// This ensures that "go list" returns a recursive list of each root
+// package's dependencies.
+func expandRecursePackages(roots []string) []string {
+	parsedRoots := []string{}
+	for _, root := range roots {
+		if strings.HasSuffix(root, "...") {
+			parsedRoots = append(parsedRoots, root)
+			continue
+		}
+
+		slash := ""
+		if string(root[len(root)-1]) != "/" {
+			slash = "/"
+		}
+		root = strings.Join([]string{root, slash, "..."}, "")
+		parsedRoots = append(parsedRoots, root)
+	}
+
+	return parsedRoots
+}
+
+func (o *TraceImportsOpts) Validate() error {
+	if len(o.Roots) == 0 {
+		return errors.New("at least one root package must be provided")
+	}
+	if len(o.OutputFormat) != 0 && o.OutputFormat != "dot" {
+		return fmt.Errorf("invalid output format provided: %s", o.OutputFormat)
+	}
+
+	return nil
+}
+
+// Run execs `go list` on all package entrypoints specified through --root.
+// Each package's ImportPath and non-transitive (immediate) imports are
+// filtered and aggregated. A package is filtered based on whether its ImportPath
+// matches an accepted pattern defined in the set of validPackages.
+// Each aggregated package becomes a node in a generated dependency graph.
+// An edge is created between a package and each of its Imports.
+func (o *TraceImportsOpts) Run() error {
+	args := []string{"list", "--json"}
+	golist := exec.Command("go", append(args, o.Roots...)...)
+
+	r, w := io.Pipe()
+	golist.Stdout = w
+	golist.Stderr = os.Stderr
+
+	pkgs := &PackageList{}
+	go func(list *PackageList) {
+		decoder := json.NewDecoder(r)
+		for {
+			var pkg Package
+			err := decoder.Decode(&pkg)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+			}
+
+			list.Add(pkg)
+		}
+	}(pkgs)
+
+	if err := golist.Run(); err != nil {
+		return err
+	}
+
+	g, nodes, err := BuildGraph(pkgs, o.Exclude)
```

Comment:
you need to wait until you've completed your decode before using the pkgs.

Reply:
the `golist.Run` call above blocks until the command finishes executing. Will add comment to clarify